cumulusci/tasks/salesforce/update_profile.py

```diff
@@ -47,168 +65,182 @@ def _init_options(self, kwargs):
-
-    def _process_metadata(self):
-        self.logger.info(f"Processing retrieved metadata in {self.retrieve_dir}")
-        path = self.retrieve_dir / "profiles" / f"{self.profile_name}.profile"
-        self.tree = elementtree_parse_file(path)
-
-        self._set_apps_visible()
-        self._set_classes_enabled()
-        self._set_fields_editable()
-        self._set_fields_readable()
-        self._set_pages_enabled()
-        self._set_tabs_visibility()
-        self._set_record_types()
-
-        self.tree.write(
-            path, "utf-8", xml_declaration=True, default_namespace=self.namespaces["sf"]
+
+    def freeze(self, step):
+        # Preserve behavior from when we subclassed Deploy.
+
+        steps = super().freeze(step)
+        for step in steps:
+            if step["kind"] == "other":
+                step["kind"] = "metadata"
+        return steps
+
+    def _generate_package_xml(self, operation):
+        if operation is MetadataOperation.RETRIEVE:
+            with open(self.package_xml_path, "r") as f:
+                package_xml_content = f.read()
+
+            package_xml_content = package_xml_content.format(**self.namespace_prefixes)
+
+            if (
+                self.options["include_packaged_objects"]
+                or "package_xml" not in self.options
+            ):
+                # We need to rewrite the package.xml for one or two reasons.
+                # Either we are using packaged-object expansion, or we're using
+                # the built-in admin_profile.xml and need to substitute in
+                # profile API names.
+                package_xml = metadata_tree.fromstring(package_xml_content)
+
+                if self.options["include_packaged_objects"]:
+                    self._expand_package_xml(package_xml)
+                if "package_xml" not in self.options:
+                    self._expand_profile_members(package_xml)
+
+                package_xml_content = package_xml.tostring(xml_declaration=True)
+
+            return package_xml_content
+        else:
+            return super()._generate_package_xml(operation)
+
+    def _expand_profile_members(self, package_xml):
+        profile_names = package_xml.find("types", name="Profile")
+        for profile in self.api_names:
+            profile_names.append("members", text=profile)
+
+    def _expand_package_xml(self, package_xml):
+        # Query the target org for all namespaced objects
+        # Add these entities to the package.xml
+
+        results = self.tooling.query(
+            "SELECT DeveloperName, NamespacePrefix FROM CustomObject WHERE ManageableState != 'unmanaged'"
         )
 
-    def _set_apps_visible(self):
-        xpath = ".//sf:applicationVisibilities[sf:visible='false']"
-        for elem in self.tree.findall(xpath, self.namespaces):
-            elem.find("sf:visible", self.namespaces).text = "true"
+        custom_objects = package_xml.find("types", name="CustomObject")
```

Comment:
If it's not there we should add it or throw a helpful exception.

Reply:
Opting for throwing an exception in both of these cases because they indicate a package.xml that is not properly configured for this task; taking action could produce unexpected results.